### PR TITLE
Fix mobile video and audio

### PR DESCRIPTION
**Pull request by blopker**

### CODE REVIEW

File: lib/app/circle/circle_join_dialog.dart [hunk 1]
```diff
@@ -227,7 +228,9 @@ class _CircleJoinDialogState extends ConsumerState<CircleJoinDialog> {
       children: [
         Stack(
           children: [
-            const rtc_local_view.SurfaceView(),
+            kIsWeb
+                ? const rtc_local_view.SurfaceView()
+                : const rtc_local_view.TextureView(),
```
Comment (blopker) [inline lib/app/circle/circle_join_dialog.dart:233]: TextureView is more performant on mobile according to the Agora devs.

File: lib/services/agora/agora_communication_provider.dart [hunk 2]
```diff
@@ -334,7 +334,16 @@ class AgoraCommunicationProvider extends CommunicationProvider {
               statusValue != PermissionStatus.limited) {
             debugPrint('Failed requesting bluetooth connect!');
           }
+          statusValue = await Permission.camera.request();
+          if (statusValue != PermissionStatus.granted &&
+              statusValue != PermissionStatus.limited) {
+            debugPrint('Failed requesting camera!');
+          }
```
Comment (blopker) [inline lib/services/agora/agora_communication_provider.dart:341]: Also ask for camera permissions up front.

File: lib/services/agora/agora_communication_provider.dart [hunk 3]
```diff
@@ -334,7 +334,16 @@ class AgoraCommunicationProvider extends CommunicationProvider {
...
           statusValue = await Permission.microphone.request();
+          if (statusValue != PermissionStatus.granted &&
+              statusValue != PermissionStatus.limited) {
+            debugPrint('Failed requesting microphone!');
+          }
```
Comment (blopker) [inline lib/services/agora/agora_communication_provider.dart:346]: This just adds a similar nice error message for microphones.

File: lib/services/agora/agora_communication_provider.dart [hunk 4]
```diff
@@ -961,14 +970,23 @@ class AgoraCommunicationProvider extends CommunicationProvider {
...
-          return audioInputDevices.isNotEmpty &&
-              audioInputDevices[0].deviceId.isNotEmpty;
+          if (kIsWeb) {
+            List<MediaDeviceInfo> audioInputDevices =
+                await _engine!.deviceManager.enumerateAudioRecordingDevices();
+            return audioInputDevices.isNotEmpty &&
+                audioInputDevices[0].deviceId.isNotEmpty;
+          } else {
+            return Permission.microphone.request().isGranted;
+          }
```
Comment (blopker) [inline lib/services/agora/agora_communication_provider.dart:989]: The core of the issue, the permission check logic was only working for web. This also has it work for mobile, optionally requesting the permission as well.

**Review by p2-kyle — APPROVED** [review]

**Pull request merged**